src/ec/eddsa.rs

```diff
@@ -133,33 +174,144 @@ impl signature::VerificationAlgorithm for EdDSAParameters {
     fn verify(&self, public_key: untrusted::Input, msg: untrusted::Input,
               signature: untrusted::Input) -> Result<(), error::Unspecified> {
         let public_key = public_key.as_slice_less_safe();
-        if public_key.len() != 32 || signature.len() != 64 {
+        if public_key.len() != PUBLIC_KEY_LEN || signature.len() != SIGNATURE_LEN {
             return Err(error::Unspecified);
         }
         let msg = msg.as_slice_less_safe();
         let signature = signature.as_slice_less_safe();
-        bssl::map_result(unsafe {
-            GFp_ed25519_verify(msg.as_ptr(), msg.len(), signature.as_ptr(),
-                               public_key.as_ptr())
-        })
+
+        let mut a = ExtPoint::new_at_infinity();
+        if (signature[63] & 224) != 0 {
+            return Err(error::Unspecified);
+        }
+        try!(bssl::map_result(unsafe {
+            GFp_x25519_ge_frombytes_vartime(&mut a, slice_as_scalar_ref(public_key).unwrap())
+        }));
+        for i in 0..ELEM_LIMBS {
+            a.x[i] = -a.x[i];
+            a.t[i] = -a.t[i];
+        }
+        let mut r_copy = [0u8; SCALAR_LEN];
+        r_copy.copy_from_slice(&signature[..SCALAR_LEN]);
+        let mut s_copy = [0u8; SCALAR_LEN];
+        s_copy.copy_from_slice(&signature[SCALAR_LEN..]);
+        let mut h = [0u8; digest::SHA512_OUTPUT_LEN];
+        h.copy_from_slice({
+            let mut ctx = digest::Context::new(&digest::SHA512);
+            ctx.update(&signature[..SCALAR_LEN]);
+            ctx.update(&public_key);
+            ctx.update(&msg);
+            ctx.finish().as_ref()
+        });
+        unsafe {
+            GFp_x25519_sc_reduce(slice_as_scalar_ref_mut(&mut h).unwrap())
+        };
+        let mut r = Point::new_at_infinity();
+        unsafe {
+            GFp_ge_double_scalarmult_vartime(&mut r,
+                                             slice_as_scalar_ref(&h).unwrap(),
+                                             &a,
+                                             slice_as_scalar_ref(&s_copy).unwrap())
+        };
+        let mut r_check = [0u8; ELEM_LEN];
+        unsafe { GFp_x25519_ge_tobytes(&mut r_check, &r) };
+        constant_time::verify_slices_are_equal(&r_check, &r_copy)
     }
 }
 
 impl private::Private for EdDSAParameters {}
 
 
+fn public_from_private(seed: &Seed, out: &mut PublicKey) {
+    let seed_sha512 = digest::digest(&digest::SHA512, seed);
+    let a_bytes = slice_as_scalar_ref(&seed_sha512.as_ref()).unwrap();
+    let mut a_bytes = *a_bytes;
+    unsafe {
+        GFp_ed25519_scalar_mask(&mut a_bytes);
+    }
+    let mut a = ExtPoint::new_at_infinity();
+    unsafe {
+        GFp_x25519_ge_scalarmult_base(&mut a, &a_bytes);
+        GFp_ge_p3_tobytes(out, &a);
+    }
+}
+
+fn slice_as_scalar_ref(slice: &[u8]) -> Result<&Scalar, error::Unspecified> {
+    slice_as_array_ref!(&slice[..SCALAR_LEN], SCALAR_LEN)
```

Comment:
Done. I've removed all use of the new functions in `sign` and `verify`, and will remove the final occurrence in `public_from_private` shortly.